core/src/main/scala/org/apache/spark/util/collection/PartitionedPairBuffer.scala

```diff
@@ -74,7 +74,20 @@ private[spark] class PartitionedPairBuffer[K, V](initialCapacity: Int = 64)
   /** Iterate through the data in a given order. For this class this is not really destructive. */
   override def partitionedDestructiveSortedIterator(keyComparator: Option[Comparator[K]])
     : Iterator[((Int, K), V)] = {
-    val comparator = keyComparator.map(partitionKeyComparator).getOrElse(partitionComparator)
+    val comparator : Comparator[(Int, K)] = 
+      if (keyComparator.isEmpty) {
+        partitionComparator
+    } else
+      new Comparator[(Int, K)] {
+        override def compare(a: (Int, K), b: (Int, K)): Int = {
+          val partitionDiff = a._1 - b._1
```

Comment:
There are some indentation problems here and the else clause is missing a brace. I think you can omit the type of `comparator`; no space before the colon in any event.

This subtraction can overflow in theory and give the wrong answer, but the existing code does it, so, pass on that.

While optimizing, do you want to call keyComparator.get outside the class definition?

There's a similar construct in PartitionedAppendOnlyMap that should be changed too. Can this be refactored maybe?

Can the method partitionKeyComparator go away? I think the whole WritablePartitionedPairCollection object goes away after this if you care to 'inline' it too in the one refactored instance.